### ISSUE TITLE
Make for/of transpiled output faster when the iterable is an array

### ISSUE THREAD

**Issue by MatrixFrog** [google/closure-compiler#1584]
In general, to transpile a for/of, we need to create an iterator, and each time through the loop, call iter.next().value to get the value. But, if we know the thing we're iterating over is an array, we can instead do a traditional `for (var i=0; i<arr.length; i++)` loop. Currently transpilation runs before typechecking so we can't know that unless it's an array literal. Someday we'll move transpilation after typechecking but it would be good to have a short term solution so that `for/of`s can be reasonably fast in the meantime (my understanding is that they are quite slow in browsers, at least for now). Maybe we add a separate optimization that runs after typechecking and knows how to remove the `iterable[Symbol.iterator]()` call and convert to a vanilla for loop.

Been meaning to figure this out for a while, thank you @robey for reminding me.

**Comment by Dominator008** [comment]
I **think** our current transpilation is already faster than a native for-of loop on V8's Crankshaft, which bails out on functions containing these loops: https://github.com/petkaantonov/bluebird/wiki/Optimization-killers#2-unsupported-syntax.

**Comment by concavelenz** [comment]
for-of, I believe is supported by turbofan.  Do you have a benchmark?

**Comment by Dominator008** [comment]
@concavelenz http://jsperf.com/native-for-of-vs-transpiled/3, I'm running Chrome 50, the transpiled version is a bit faster.

**Comment by concavelenz** [comment]
Right.  We would like folks to just write the natural thing, not hand written loops.
